// hono/dist/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Context = void 0;
const url_1 = require("./utils/url");
class Context {
    constructor(req, env = undefined, event = undefined, notFoundHandler = () => new Response()) {
        this._status = 200;
        this._pretty = false;
        this._prettySpace = 2;
        this.req = req;
        if (env) {
            this.env = env;
        }
        this.event = event;
        this.notFoundHandler = notFoundHandler;
        this.finalized = false;
    }
    get res() {
        return (this._res || (this._res = new Response()));
    }
    set res(_res) {
        this._res = _res;
        this.finalized = true;
    }
    header(name, value) {
        this._headers || (this._headers = {});
        this._headers[name] = value;
        if (this.finalized) {
            this.res.headers.set(name, value);
        }
    }
    status(status) {
        this._status = status;
    }
    set(key, value) {
        this._map || (this._map = {});
        this._map[key] = value;
    }
    get(key) {
        if (!this._map) {
            return undefined;
        }
        return this._map[key];
    }
    pretty(prettyJSON, space = 2) {
        this._pretty = prettyJSON;
        this._prettySpace = space;
    }
    newResponse(data, status, headers = {}) {
        const _headers = { ...this._headers, ...headers };
        if (this._res) {
            this._res.headers.forEach((v, k) => {
                _headers[k] = v;
            });
        }
        return new Response(data, {
            status: status || this._status || 200,
            headers: _headers,
        });
    }
    body(data, status = this._status, headers = {}) {
        return this.newResponse(data, status, headers);
    }
    text(text, status = this._status, headers = {}) {
        headers['Content-Type'] || (headers['Content-Type'] = 'text/plain; charset=UTF-8');
        return this.body(text, status, headers);
    }
    json(object, status = this._status, headers = {}) {
        const body = this._pretty
            ? JSON.stringify(object, null, this._prettySpace)
            : JSON.stringify(object);
        headers['Content-Type'] || (headers['Content-Type'] = 'application/json; charset=UTF-8');
        return this.body(body, status, headers);
    }
    html(html, status = this._status, headers = {}) {
        headers['Content-Type'] || (headers['Content-Type'] = 'text/html; charset=UTF-8');
        return this.body(html, status, headers);
    }
    redirect(location, status = 302) {
        if (!(0, url_1.isAbsoluteURL)(location)) {
            const url = new URL(this.req.url);
            url.pathname = location;
            location = url.toString();
        }
        return this.newResponse(null, status, {
            Location: location,
        });
    }
    notFound() {
        return this.notFoundHandler(this);
    }
}
exports.Context = Context;
